### ISSUE TITLE
Game clipboard is not shared with system clipboard when running HTML5 exports

### ISSUE THREAD

**Issue by Goutte** [godotengine/godot#12587]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**

- Debian and Ubuntu, on Chromium and Firefox.
- Godot 2.1.4 stable
- HTML5 exports only (work with regular binaries)

**Issue description:**

System clipboard contents cannot be pasted in controls such as `LineEdit`. The game has its own clipboard contents and it is not shared with the system.

`OS.get_clipboard()` stays empty unless we copy text from the game itself.

This is confusing for users because HTML `input` fields do share their clipboard with the system.

**Steps to reproduce:**

1. Make a scene with a `LineEdit`.
2. Export it to HTML5
3. Run it in a browser (I use `php -S` when I'm in a hurry or too lazy, to avoid `file://` issues)
4. Copy text from any source but the game (eg: the URL bar of the browser)
5. Try to paste it in the `LineEdit`.

**Comment by Zylann** [comment]
Isn't this a security concern? Maybe HTML input fields do share input (because they are handled by the browser itself, not the web app), but an exported Godot game is seen like a Javascript web app AFAIK, which should not be able to just look at your clipboard, login name, filesystem or stuff like this without consent. For this reason it would even be possible that, even if working, this feature would be limited to specific cases or would pop a window on the user's browser before returning anything (which would freeze the game as it is now btw).

**Comment by Goutte** [comment]
Thanks @Zylann for the link and the concern!

Clipboard access has been granted to web apps by browser makers, and a godot exported game is a web app like any other. Why should it deprive itself of a feature that the rest of the web enjoys?
It is a privacy concern, but clipboard contents have always been considered fair game because it's their very ubiquitousness that makes them useful.
---
I'm not sure how to implement that feature (or fix that bug, whatever), nor where the related code is. It is a critical component of my game, though, and it makes me a sad panda. Any help would be appreciated.

**Comment by Zylann** [comment]
@Goutte well, I'm not saying it should be the way I described, but the resources I looked up by doing a search imply that it actually IS like that. But I could be wrong, we are in 2017 now :p

**Comment by Goutte** [comment]
I just tested the Selection API in a fiddle, and it seems to be working without prompting for any permissions. Good for us!

---

Inquiring furthermore, I found an [interesting byt of code here](https://github.com/godotengine/godot/blob/2bece6bbd35c53390b2d3eedcf9b6d9949db518d/platform/javascript/os_javascript.cpp#L659).

Maybe @reduz can help us ? It's been four years, though.

**Comment by eska014** [comment]
Many of the comments in `OS_JavaScript`, like that one, are remnants from `OS_Android`, from which that implementation was copied as a base.

The issue with clipboard sync is pasting. Godot needs the OS clipboard data during the `keydown` event, but we only get the that data in the `paste` event, which occurs after the `keydown` event. So it has to be hacked. Selection API doesn't apply to us because we render text manually using FreeType + WebGL

**Comment by Goutte** [comment]
Would that help if support for the `paste` event was added to emscripten ?
I can't find it in [its list of events](https://kripken.github.io/emscripten-site/docs/api_reference/html5.h.html#ui), nor in its issues.

Could we figure out the currently focused control node from `os_javascript.cpp`, if emscripten had paste support ?

**Comment by eska014** [comment]
No need for Emscripten APIs, JavaScript can be used directly, similar as already done for `blur`, `focus`, `mouseover` and `mouseleave`.
The currently focused Control receives the Ctrl+V input event in its `_gui_input` callback and handles pasting there if the node class allows text input.

**Comment by Goutte** [comment]
Thanks so much for the invaluable help, @eska014 !

---

Would it be appropriate to trigger a `MainLoop::NOTIFICATION_WM_PASTE` in the fashion of `mouseleave` and others ?
Since such notifications don't seem to be able to transport parameters, I'm not sure.

So far, I've got :
1. [Listen to](https://github.com/godotengine/godot/blob/7715a261d5f387b7769bb8149735e4131ea97757/platform/javascript/os_javascript.cpp#L806) the `paste` event
2. In that listener, get the pasted string
3. Store that pasted string somewhere
4. Trigger an internal event (notification?)
5. Hack something [around here](https://github.com/godotengine/godot/blob/f2f75d1f117e52f3f4a033fdf967d35bb3dbaa42/scene/main/viewport.cpp#L1439) maybe
6. ???
7. Publish!

**Comment by eska014** [comment]
There's no need to implement copy/paste, this is implemented per Control class in `_gui_input`. What needs to happen is a call to `OS::set_clipboard`, passing in the data of the actual operating system's clipboard, overriding Godot's internal clipboard. This needs to happen before `_gui_input` is called.

The actual operating system's clipboard can be read from the `paste` event, but it occurs too late.
Additionally, that event is only fired when hitting Ctrl+V, so it can't work for *right&nbsp;click*&nbsp;→&nbsp;*Paste* either.

`document.execCommand('paste')` might work, but is only available to WebExtensions.

The new Async Clipboard API might work, but isn't available yet.

**Comment by akien-mga** [comment]
This is partially addressed in #29298, but there are some limitations as described in that PR.

**Comment by bfishman** [comment]
Sorry to necro this issue, particularly because I have no code to contribute. Just looking to bump this issue, noting that now with the focus on web export for the in-browser editor feature, it may be an appropriate time to readdress the clipboard issue.

**Comment by Calinou** [comment]
@bfishman I'm not sure if it can be fixed completely, given how browsers have hardened their clipboard security recently.

**Comment by Faless** [comment]
@bfishman did you see #29298 linked above? Clipboard support in HTML5 is currently implemented the best we can on browser that provides that feature. There isn't much we can do to improve that as of now.

**Comment by akien-mga** [comment]
@Faless Should we close this as "fixed as best we can" then?

**Comment by Faless** [comment]
> @Faless Should we close this as "fixed as best we can" then?

Yeah, I think we should, I thought there was room for improvements in `4.0` since deferred input reporting would fix the first-paste issue, but that causes a whole set of new problems (most `OS` actions are only allowed during DOM input callbacks).

Closing.

**Comment by bfishman** [comment]
@Faless thanks for the reply. I'll explore implementing my own 'hack' to keep the engine's clipboard in sync with the OS. 

I don't really understand the intricacies of this yet so pardon the potentially naive question --  do you think it's possible to trigger a copy or paste event using an in-engine GUI button (as opposed to ctrl+v or ctrl)?  Obviously not a global solution but it would be enough of a workaround for my use case (many of the potential users aren't power users and may actually find a copy/paste button more useful than the normal shortcuts)

**Comment by Faless** [comment]
>  do you think it's possible to trigger a copy or paste event using an in-engine GUI button

For browser that supports the Clipboard API `readText` function (chrome) yes. Just call `OS.get_clipboard()` during the pressed signal of your button.

**Comment by GithubThrowaway69** [comment]
@Faless 
Hey guys I'm a little confused by reading all of this. Is it possible to allow users to paste text into an HTML5 TextEdit using CTRL + V?

I have been trying to detect the pressing of CTRL + V and then read from the javascript clipboard but have not had luck. Let me know if this is actually impossible so I can stop trying lmao.

I do understand this can work in browsers like chrome by granting permissions, but I would like users to be able to paste with just CTRL + V so they hopefully don't need to grant permissions and it would work in Firefox.

**Comment by Faless** [comment]
> I have been trying to detect the pressing of CTRL + V with an input event and then read from the javascript clipboard but have not had luck. Let me know if this is actually impossible so I can stop trying lmao.

Not really, due to the order in which DOM events are fired, we need to handle the CTRL (or COMMAND on Mac) and V keys separately, and prevent their default (to avoid browsers using those input in other ways), and the paste event is sent after the CTRL and V, and only if those keys are not being prevented their default behaviour.
Your best bet, is to tell the users to use the "Edit -> Paste" menu entry in their browser. That should always work (though I agree is really bad UX).

**Comment by me2beats** [comment]
> Your best bet, is to tell the users to use the "Edit -> Paste" menu entry in their browser. That should always work (though I agree is really bad UX).

How to make them work?
They don't work with Godot text input controls

**Comment by Patashu** [comment]
For future people who click on this issue and don't need a LOT of clipboard access, here is my workaround that I'll be using for my own Godot HTML5 games using the 'prompt' function: https://www.reddit.com/r/godot/comments/17rsemr/how_to_access_the_clipboard_as_a_html5web_godot/

I hope it helps someone out there!

**Comment by Calinou** [comment]
> For future people who click on this issue and don't need a LOT of clipboard access, here is my workaround that I'll be using for my own Godot HTML5 games using the 'prompt' function: [reddit.com/r/godot/comments/17rsemr/how_to_access_the_clipboard_as_a_html5web_godot](https://www.reddit.com/r/godot/comments/17rsemr/how_to_access_the_clipboard_as_a_html5web_godot/)
> 
> I hope it helps someone out there!

This could be interesting to expose as an export preset option (something like **Emulate Clipboard**).

**Comment by Patashu** [comment]
I got another suggestion which works very well, but I'm not sure how easy it'd be to implement in Godot, because it would effectively make get_clipboard/set_clipboard asynchronous or have to pause the game: "for N Step Steve, I made some HTML elements which are hidden most of the time but shown when the player needs to input a level code."

I think people who want copying/pasting directly to LineEdits are SOL, though, unless Godot is modified to make every LineEdit an HTML textbox, which haha I understand why that won't happen.
